core/src/main/java/io/questdb/griffin/engine/functions/groupby/ApproxPercentileDoubleGroupByFunction.java

```diff
@@ -0,0 +1,198 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.SymbolTableSource;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.*;
+import io.questdb.std.Numbers;
+import io.questdb.std.ObjList;
+import io.questdb.std.histogram.org.HdrHistogram.DoubleHistogram;
+
+public class ApproxPercentileDoubleGroupByFunction extends DoubleFunction implements GroupByFunction, TernaryFunction {
+    private final int funcPosition;
+    private final Function exprFunc;
+    private final Function percentileFunc;
+    // specifies the precision for the recorded values (between 0 and 5, defaults to 3).
+    // trade-off between memory usage and accuracy.
+    private final Function precisionFunc;
+    private final ObjList<DoubleHistogram> histograms = new ObjList<>();
+    private int histogramIndex;
+    private int valueIndex;
+
+    Function defaultPrecisionFunc = new IntFunction() {
```

Comment:
This function looks unnecessary. It's simpler to get and validate the `int` value in the `ApproxPercentileDoubleGroupByFunctionFactory` class and then pass it here.

Reply:
Right, I was overthinking this part. 

Passing it as an int makes sense. Perhaps we should do the same for `percentile` too? Since it is a constant as well.

Reply:
Never mind, I just remembered since we allow bind variables too, we can't do the same for percentile.

Reply:
I'd better keep bind variable support for the `percentile` argument, so it can be runtime constant as well. To me, this may be valuable for some users who want to, say, expose the percentile in their app UI and then sent it as a bind variable instead of ugly query string concatentation. WDYT?

Reply:
Yeah, i think it makes sense to support bind variables for `percentile`, so let's keep it that way.